File: JsonInflatorIOS/JsonInflator/Classes/JsonInflator.swift
//
//  JsonInflator.swift
//  Json inflator Pod
//
//  Library: json inflator
//  The main interface to register and inflate new objects
//

fileprivate class JsonInflatableItem {

    var inflatable: JsonInflatable?
    var attributeSets = [String: [String: Any]]()

}

open class JsonInflator {
    
    // --
    // MARK: Singleton instance
    // --

    static let shared = JsonInflator(inflatableKey: "inflator", attributeSetKey: "attributeSet")
    
    
    // --
    // MARK: Members
    // --

    public var convUtil = InflatorConvUtil()
    private let inflatableKey: String
    private let attributeSetKey: String
    private var registeredInflators = [String: JsonInflatableItem]()
    private var mergeSubAttributes = [String]()
    private var excludeAttributes = [String]()

    public var colorLookup: InflatorColorLookup? {
        didSet {
            convUtil = InflatorConvUtil(colorLookup: colorLookup, dimensionLookup: dimensionLookup)
        }
    }
    
    public var dimensionLookup: InflatorDimensionLookup? {
        didSet {
            convUtil = InflatorConvUtil(colorLookup: colorLookup, dimensionLookup: dimensionLookup)
        }
    }
    

    // --
    // MARK: Initialization
    // --
    
    public init(inflatableKey: String, attributeSetKey: String = "attributeSet") {
        self.inflatableKey = inflatableKey
        self.attributeSetKey = attributeSetKey
    }


    // --
    // MARK: Attribute inclusion/exclusion
    // --
    
    public func setMergeSubAttributes(_ attributeNames: [String]) {
        mergeSubAttributes = attributeNames
    }
    
    public func setExcludeAttributes(_ attributeNames: [String]) {
        excludeAttributes = attributeNames
    }
    
    
    // --
    // MARK: Inflatable registry
    // --
    
    public func register(name: String, inflatable: JsonInflatable) {
        obtainInflator(name).inflatable = inflatable
    }
    
    public func registerAttributeSet(inflatableName: String, setName: String, setAttributes: [String: Any]?) {
        let inflator = obtainInflator(inflatableName)
        if let setAttributes = setAttributes {
            inflator.attributeSets[setName] = setAttributes
        } else {
            let _ = inflator.attributeSets.removeValue(forKey: setName)
        }
    }
    
    public func registeredInflatableNames() -> [String] {
        return registeredInflators.keys.map { $0 }
    }
    
    private func obtainInflator(_ name: String) -> JsonInflatableItem {
        if let inflator = registeredInflators[name] {
            return inflator
        }
        let addItem = JsonInflatableItem()
        registeredInflators[name] = addItem
        return addItem
    }


    // --
    // MARK: Create and update
    // --

    public func inflate(attributes: [String: Any], parent: Any? = nil, binder: InflatorBinder? = nil) -> Any? {
        if let inflatable = findInflatableInAttributes(attributes), let inflatableName = findInflatableNameInAttributes(attributes) {
            let object = inflatable.create()
            if let mergedAttributes = processedAttributes(given: attributes, fallback: attributesForSet(inflatableName: inflatableName, setName: attributes[attributeSetKey] as? String), mergeSubAttributes: mergeSubAttributes, excludeAttributes: excludeAttributes) {
                inflatable.update(convUtil: convUtil, object: object, attributes: mergedAttributes, parent: parent, binder: binder)
            }
            return object
        }
        return nil
    }
    
    @discardableResult
    public func inflate(onObject: Any, attributes: [String: Any]?, parent: Any? = nil, binder: InflatorBinder? = nil) -> Bool {
        if attributes == nil {
            return false
        }
        if let inflatable = findInflatableInAttributes(attributes ?? [:]), let inflatableName = findInflatableNameInAttributes(attributes ?? [:]) {
            if let mergedAttributes = processedAttributes(given: attributes, fallback: attributesForSet(inflatableName: inflatableName, setName: attributes?[attributeSetKey] as? String), mergeSubAttributes: mergeSubAttributes, excludeAttributes: excludeAttributes) {
                return inflatable.update(convUtil: convUtil, object: onObject, attributes: mergedAttributes, parent: parent, binder: binder)
            }
        }
        return false
    }
    
    public func canRecycle(object: Any?, attributes: [String: Any]?) -> Bool {
        if let object = object {
            if let inflatable = findInflatableInAttributes(attributes ?? [:]) {
                return inflatable.canRecycle(convUtil: convUtil, object: object, attributes: attributes ?? [:])
            }
        }
        return false
    }
    
    public func findInflatableInAttributes(_ attributes: [String: Any]) -> JsonInflatable? {
        if let inflatorName = attributes[inflatableKey] as? String {
            return registeredInflators[inflatorName]?.inflatable
        }
        return nil
    }
    
    public func findInflatableNameInAttributes(_ attributes: [String: Any]) -> String? {
        return attributes[inflatableKey] as? String
    }
    
    
    // --
    // MARK: Nested inflation utilities
    // --
    
    public func inflateNestedItem(currentItem: Any, newItem: Any?, enableRecycling: Bool, parent: Any? = nil, binder: InflatorBinder?) -> InflatorNestedResult {
        // Recycle or inflate new item
        let result = InflatorNestedResult()
        let processedNewItem = attributesForNestedInflatable(newItem)
        var inflatedItem: Any?
        if enableRecycling && canRecycle(object: currentItem, attributes: processedNewItem) {
            inflate(onObject: currentItem, attributes: processedNewItem, parent: parent, binder: binder)
            inflatedItem = currentItem
        } else {
            result.addRemovedItem(currentItem)
            if let processedNewItem = processedNewItem {
                inflatedItem = inflate(attributes: processedNewItem, parent: parent, binder: binder)
            }
        }
        
        // Add item to result
        if let inflatedItem = inflatedItem, let newItem = processedNewItem {
            result.addItem(inflatedItem, attributes: newItem, recycled: inflatedItem as AnyObject === currentItem as AnyObject)
        }
        return result
    }

    public func inflateNestedItems(currentItems: [Any], newItems: Any?, enableRecycling: Bool, parent: Any? = nil, binder: InflatorBinder?) -> InflatorNestedResult {
        let result = InflatorNestedResult()
        let processedNewItems = attributesForNestedInflatableList(newItems)
        if enableRecycling {
            // Add or recycle items
            var recycleIndex = 0
            for newItem in processedNewItems {
                // Search for a current item to recycle (use an index to maintain order)
                var recycled = false
                var inflatedItem: Any?
                for index in recycleIndex..<currentItems.count {
                    if canRecycle(object: currentItems[index], attributes: newItem) {
                        for removeIndex in recycleIndex..<index {
                            result.addRemovedItem(currentItems[removeIndex])
                        }
                        recycleIndex = index + 1
                        inflatedItem = currentItems[index]
                        inflate(onObject: currentItems[index], attributes: newItem, parent: parent, binder: binder)
                        recycled = true
                        break
                    }
                }
                
                // If no candidate was found, create a new item
                if !recycled {
                    inflatedItem = inflate(attributes: newItem, parent: parent, binder: binder)
                }
                if let inflatedItem = inflatedItem {
                    result.addItem(inflatedItem, attributes: newItem, recycled: recycled)
                }
            }
            
            // Set remaining items for removal
            for index in recycleIndex..<currentItems.count {
                result.addRemovedItem(currentItems[index])
            }
        } else {
            // First mark all current items as removed
            for item in currentItems {
                result.addRemovedItem(item)
            }
            
            // Create new items
            for newItem in processedNewItems {
                if let inflatedItem = inflate(attributes: newItem, parent: parent, binder: binder) {
                    result.addItem(inflatedItem, attributes: newItem, recycled: false)
                }
            }
        }
        return result
    }

    public func attributesForNestedInflatable(_ nestedInflatableItem: Any?) -> [String: Any]? {
        if let attributes = nestedInflatableItem as? [String: Any] {
            if let inflatableName = findInflatableNameInAttributes(attributes) {
                return processedAttributes(given: attributes, fallback: attributesForSet(inflatableName: inflatableName, setName: attributes[attributeSetKey] as? String), mergeSubAttributes: mergeSubAttributes, excludeAttributes: excludeAttributes)
            }
        }
        return nil
    }

    public func attributesForNestedInflatableList(_ nestedInflatableItemList: Any?) -> [[String: Any]] {
        var inflatableItemList: [[String: Any]] = []
        if let itemList = nestedInflatableItemList as? [[String: Any]] {
            for item in itemList {
                if let inflatableName = findInflatableNameInAttributes(item) {
                    if let resultAttributes = processedAttributes(given: item, fallback: attributesForSet(inflatableName: inflatableName, setName: item[attributeSetKey] as? String), mergeSubAttributes: mergeSubAttributes, excludeAttributes: excludeAttributes) {
                        inflatableItemList.append(resultAttributes)
                    }
                }
            }
        }
        return inflatableItemList
    }


    // --
    // MARK: Attribute processing
    // --
    
    private func attributesForSet(inflatableName: String, setName: String?) -> [String: Any]? {
        if let inflator = registeredInflators[inflatableName] {
            if setName == "default" {
                return inflator.attributeSets["default"]
            }
            return mergedAttributes(given: inflator.attributeSets[setName ?? ""], fallback: inflator.attributeSets["default"])
        }
        return nil
    }
    
    private func processedAttributes(given: [String: Any]?, fallback: [String: Any]?, mergeSubAttributes: [String], excludeAttributes: [String]) -> [String: Any]? {
        if var result = mergedAttributes(given: given, fallback: fallback) {
            for mergeSubAttribute in mergeSubAttributes {
                if let item = result[mergeSubAttribute] as? [String: Any] {
                    result = mergedAttributes(given: item, fallback: result) ?? result
                }
            }
            for excludeAttribute in excludeAttributes {
                result.removeValue(forKey: excludeAttribute)
            }
            return result
        }
        return nil
    }
    
    private func mergedAttributes(given: [String: Any]?, fallback: [String: Any]?) -> [String: Any]? {
        // Just return one of the attributes if the other is null
        if fallback == nil {
            return given
        } else if given == nil {
            return fallback
        }
        
        // Merge and return without modifying the originals
        var merged: [String: Any] = [:]
        for (key, value) in given! {
            merged[key] = value
        }
        for (key, value) in fallback! {
            if merged[key] == nil {
                merged[key] = value
            }
        }
        return merged
    }

}

File: JsonInflatorAndroid/JsonInflatorLib/src/main/java/com/crescentflare/jsoninflator/JsonInflator.java
package com.crescentflare.jsoninflator;

import android.content.Context;

import com.crescentflare.jsoninflator.binder.InflatorBinder;
import com.crescentflare.jsoninflator.utility.InflatorColorLookup;
import com.crescentflare.jsoninflator.utility.InflatorDimensionLookup;
import com.crescentflare.jsoninflator.utility.InflatorMapUtil;
import com.crescentflare.jsoninflator.utility.InflatorNestedResult;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Json inflator: manages inflatables
 * The main interface to register and inflate new objects
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class JsonInflator
{
    // --
    // Singleton instance
    // --

    public static JsonInflator instance = new JsonInflator("inflator", "attributeSet");


    // --
    // Members
    // --

    private Map<String, InflatableItem> registeredInflators = new HashMap<>();
    private List<String> mergeSubAttributes = new ArrayList<>();
    private List<String> excludeAttributes = new ArrayList<>();
    private InflatorMapUtil mapUtil = new InflatorMapUtil();
    private InflatorColorLookup colorLookup = null;
    private InflatorDimensionLookup dimensionLookup = null;
    private String inflatableKey;
    private String attributeSetKey;


    // --
    // Initialization
    // --

    public JsonInflator(@NotNull String inflatableKey)
    {
        this(inflatableKey, "attributeSet");
    }

    public JsonInflator(@NotNull String inflatableKey, @NotNull String attributeSetKey)
    {
        this.inflatableKey = inflatableKey;
        this.attributeSetKey = attributeSetKey;
    }


    // --
    // Access conversion properties
    // --

    @NotNull
    public InflatorMapUtil getMapUtil()
    {
        return mapUtil;
    }

    public void setColorLookup(@Nullable InflatorColorLookup colorLookup)
    {
        this.colorLookup = colorLookup;
        mapUtil = new InflatorMapUtil(colorLookup, dimensionLookup);
    }

    public void setDimensionLookup(@Nullable InflatorDimensionLookup dimensionLookup)
    {
        this.dimensionLookup = dimensionLookup;
        mapUtil = new InflatorMapUtil(colorLookup, dimensionLookup);
    }


    // --
    // Attribute inclusion/exclusion
    // --

    public void setMergeSubAttributes(@NotNull List<String> attributeNames)
    {
        mergeSubAttributes = attributeNames;
    }

    public void setExcludeAttributes(@NotNull List<String> attributeNames)
    {
        excludeAttributes = attributeNames;
    }


    // --
    // Inflatable registry
    // --

    public void register(@NotNull String name, @Nullable JsonInflatable inflatable)
    {
        obtainInflator(name).setInflatable(inflatable);
    }

    public void registerAttributeSet(@NotNull String inflatableName, @NotNull String setName, @Nullable Map<String, Object> setAttributes)
    {
        InflatableItem inflator = obtainInflator(inflatableName);
        if (setAttributes != null)
        {
            inflator.getAttributeSets().put(setName, setAttributes);
        }
        else
        {
            inflator.getAttributeSets().remove(setName);
        }
    }

    @NotNull
    public List<String> registeredInflatableNames()
    {
        return new ArrayList<>(registeredInflators.keySet());
    }

    private InflatableItem obtainInflator(@NotNull String name)
    {
        InflatableItem item = registeredInflators.get(name);
        if (item == null)
        {
            item = new InflatableItem();
            registeredInflators.put(name, item);
        }
        return item;
    }


    // --
    // Create and update
    // --

    @Nullable
    public Object inflate(@NotNull Context context, @Nullable Map<String, Object> attributes, @Nullable Object parent)
    {
        return inflate(context, attributes, parent, null);
    }

    @Nullable
    public Object inflate(@NotNull Context context, @Nullable Map<String, Object> attributes, @Nullable Object parent, @Nullable InflatorBinder binder)
    {
        String inflatableName = findInflatableNameInAttributes(attributes);
        if (inflatableName != null)
        {
            JsonInflatable inflatable = findInflatableInAttributes(attributes);
            if (inflatable != null)
            {
                Object object = inflatable.create(context);
                Map<String, Object> mergedAttributes = processAttributes(attributes, attributesForSet(inflatableName, mapUtil.optionalString(attributes, attributeSetKey, null)), mergeSubAttributes, excludeAttributes);
                inflatable.update(mapUtil, object, mergedAttributes, parent, binder);
                return object;
            }
        }
        return null;
    }

    public boolean inflateOn(@NotNull Object object, @Nullable Map<String, Object> attributes, @Nullable Object parent)
    {
        return inflateOn(object, attributes, parent, null);
    }

    public boolean inflateOn(@NotNull Object object, @Nullable Map<String, Object> attributes, @Nullable Object parent, @Nullable InflatorBinder binder)
    {
        String inflatableName = findInflatableNameInAttributes(attributes);
        if (inflatableName != null)
        {
            JsonInflatable inflatable = findInflatableInAttributes(attributes);
            if (inflatable != null)
            {
                Map<String, Object> mergedAttributes = processAttributes(attributes, attributesForSet(inflatableName, mapUtil.optionalString(attributes, attributeSetKey, null)), mergeSubAttributes, excludeAttributes);
                return inflatable.update(mapUtil, object, mergedAttributes, parent, binder);
            }
        }
        return false;
    }

    public boolean canRecycle(@Nullable Object object, @Nullable Map<String, Object> attributes)
    {
        if (object != null && attributes != null)
        {
            JsonInflatable inflatable = findInflatableInAttributes(attributes);
            if (inflatable != null)
            {
                return inflatable.canRecycle(mapUtil, object, attributes);
            }
        }
        return false;
    }

    @Nullable
    public JsonInflatable findInflatableInAttributes(@Nullable Map<String, Object> attributes)
    {
        String inflatorName = mapUtil.optionalString(attributes, inflatableKey, null);
        if (inflatorName != null)
        {
            InflatableItem item = registeredInflators.get(inflatorName);
            if (item != null)
            {
                return item.inflatable;
            }
        }
        return null;
    }

    @Nullable
    public String findInflatableNameInAttributes(@Nullable Map<String, Object> attributes)
    {
        return mapUtil.optionalString(attributes, inflatableKey, null);
    }


    // --
    // Nested inflation utilities
    // --

    @NotNull
    public InflatorNestedResult inflateNestedItem(@NotNull Context context, @NotNull Object currentItem, @Nullable Object newItem, boolean enableRecycling, @Nullable Object parent, @Nullable InflatorBinder binder)
    {
        // Recycle or inflate new item
        InflatorNestedResult result = new InflatorNestedResult();
        Map<String, Object> processedNewItem = attributesForNestedInflatable(newItem);
        Object inflatedItem;
        if (enableRecycling && canRecycle(currentItem, processedNewItem))
        {
            inflateOn(currentItem, processedNewItem, parent, binder);
            inflatedItem = currentItem;
        }
        else
        {
            result.addRemovedItem(currentItem);
            inflatedItem = inflate(context, processedNewItem, parent, binder);
        }

        // Add item to result
        if (inflatedItem != null && processedNewItem != null)
        {
            result.addItem(inflatedItem, processedNewItem, inflatedItem == currentItem);
        }
        return result;
    }

    @NotNull
    public InflatorNestedResult inflateNestedItemList(@NotNull Context context, List<Object> currentItems, @Nullable Object newItems, boolean enableRecycling, @Nullable Object parent, @Nullable InflatorBinder binder)
    {
        InflatorNestedResult result = new InflatorNestedResult();
        List<Map<String, Object>> processedNewItems = attributesForNestedInflatableList(newItems);
        if (enableRecycling)
        {
            // Add or recycle items
            int recycleIndex = 0;
            for (Map<String, Object> newItem : processedNewItems)
            {
                // Search for a current item to recycle (use an index to maintain ordering)
                boolean recycled = false;
                Object inflatedItem = null;
                for (int index = recycleIndex; index < currentItems.size(); index++)
                {
                    if (canRecycle(currentItems.get(index), newItem))
                    {
                        for (int removeIndex = recycleIndex; removeIndex < index; removeIndex++)
                        {
                            result.addRemovedItem(currentItems.get(removeIndex));
                        }
                        recycleIndex = index + 1;
                        inflatedItem = currentItems.get(index);
                        inflateOn(currentItems.get(index), newItem, parent, binder);
                        recycled = true;
                        break;
                    }
                }

                // If no candidate was found, create a new item
                if (!recycled)
                {
                    inflatedItem = inflate(context, newItem, parent, binder);
                }
                if (inflatedItem != null)
                {
                    result.addItem(inflatedItem, newItem, recycled);
                }
            }

            // Set remaining items for removal
            for (int index = recycleIndex; index < currentItems.size(); index++)
            {
                result.addRemovedItem(currentItems.get(index));
            }
        }
        else
        {
            // First mark all current items as removed
            for (Object item : currentItems)
            {
                result.addRemovedItem(item);
            }

            // Create new items
            for (Map<String, Object> newItem : processedNewItems)
            {
                Object inflatedItem = inflate(context, newItem, parent, binder);
                if (inflatedItem != null)
                {
                    result.addItem(inflatedItem, newItem, false);
                }
            }
        }
        return result;
    }

    @Nullable
    public Map<String, Object> attributesForNestedInflatable(@Nullable Object nestedInflatableItem)
    {
        Map<String, Object> attributes = mapUtil.asStringObjectMap(nestedInflatableItem);
        if (attributes != null)
        {
            String inflatableName = findInflatableNameInAttributes(attributes);
            if (inflatableName != null)
            {
                return processAttributes(attributes, attributesForSet(inflatableName, mapUtil.optionalString(attributes, attributeSetKey, null)), mergeSubAttributes, excludeAttributes);
            }
        }
        return null;
    }

    public List<Map<String, Object>> attributesForNestedInflatableList(@Nullable Object nestedInflatableItemList)
    {
        List<Map<String, Object>> inflatableItemList = new ArrayList<>();
        if (nestedInflatableItemList instanceof List<?>)
        {
            List<?> itemList = (List<?>)nestedInflatableItemList;
            for (Object item : itemList)
            {
                Map<String, Object> attributes = mapUtil.asStringObjectMap(item);
                if (attributes != null)
                {
                    String inflatableName = findInflatableNameInAttributes(attributes);
                    if (inflatableName != null)
                    {
                        inflatableItemList.add(processAttributes(attributes, attributesForSet(inflatableName, mapUtil.optionalString(attributes, attributeSetKey, null)), mergeSubAttributes, excludeAttributes));
                    }
                }
            }
        }
        return inflatableItemList;
    }


    // --
    // Attribute processing
    // --

    @Nullable
    private Map<String, Object> attributesForSet(@NotNull String inflatableName, @Nullable String setName)
    {
        InflatableItem inflator = registeredInflators.get(inflatableName);
        if (inflator != null)
        {
            String checkSetName = setName != null ? setName : "";
            if (checkSetName.equals("default"))
            {
                return inflator.getAttributeSets().get("default");
            }
            return mergeAttributes(inflator.getAttributeSets().get(setName), inflator.getAttributeSets().get("default"));
        }
        return null;
    }

    @NotNull
    private Map<String, Object> processAttributes(@Nullable Map<String, Object> givenAttributes, @Nullable Map<String, Object> fallbackAttributes, @NotNull List<String> mergeSubAttributes, @NotNull List<String> excludeAttributes)
    {
        Map<String, Object> result = mergeAttributes(givenAttributes, fallbackAttributes);
        for (String mergeSubAttribute : mergeSubAttributes)
        {
            Map<String, Object> item = mapUtil.asStringObjectMap(result.get(mergeSubAttribute));
            if (item != null)
            {
                result = mergeAttributes(item, result);
            }
        }
        for (String excludeAttribute : excludeAttributes)
        {
            result.remove(excludeAttribute);
        }
        return result;
    }

    @NotNull
    private static Map<String, Object> mergeAttributes(@Nullable Map<String, Object> givenAttributes, @Nullable Map<String, Object> fallbackAttributes)
    {
        // Just return one of the attributes if the other is null
        if (fallbackAttributes == null)
        {
            return givenAttributes != null ? givenAttributes : new HashMap<String, Object>();
        }
        else if (givenAttributes == null)
        {
            return fallbackAttributes;
        }

        // Merge and return without modifying the originals
        Map<String, Object> mergedAttributes = new HashMap<>();
        for (String key : givenAttributes.keySet())
        {
            Object addItem = givenAttributes.get(key);
            if (addItem != null)
            {
                mergedAttributes.put(key, addItem);
            }
        }
        for (String key : fallbackAttributes.keySet())
        {
            if (!mergedAttributes.containsKey(key))
            {
                Object addItem = fallbackAttributes.get(key);
                if (addItem != null)
                {
                    mergedAttributes.put(key, addItem);
                }
            }
        }
        return mergedAttributes;
    }


    // --
    // Helper class to group inflatable and its attribute sets
    // --

    private class InflatableItem
    {
        private JsonInflatable inflatable;
        private Map<String, Map<String, Object>> attributeSets = new HashMap<>();

        public JsonInflatable getInflatable()
        {
            return inflatable;
        }

        public void setInflatable(JsonInflatable inflatable)
        {
            this.inflatable = inflatable;
        }

        public Map<String, Map<String, Object>> getAttributeSets()
        {
            return attributeSets;
        }

        public void setAttributeSets(Map<String, Map<String, Object>> attributeSets)
        {
            this.attributeSets = attributeSets;
        }
    }
}
